File: VB/WinWebSolution.Module/Documents.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Validation

Namespace WinWebSolution.Module
	<DefaultClassOptions, System.ComponentModel.DisplayName("Documents"), DefaultProperty("Name")> _
	Public MustInherit Class DocumentBase
		Inherits Note
		Protected Const AssignedToContactCriteriaBase As String = "Active=True"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		<VisibleInDetailView(False), VisibleInListView(False), VisibleInLookupListView(False)> _
		Public Shadows Property Author() As String
			Get
				Return MyBase.Author
			End Get
			Set(ByVal value As String)
				MyBase.Author = value
			End Set
		End Property
		Private nameCore As String
		<RuleRequiredField(Nothing, DefaultContexts.Save)> _
		Public Property Name() As String
			Get
				Return nameCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Name", nameCore, value)
			End Set
		End Property
		Private assignedToCore As ContactBase
		<DataSourceCriteria(AssignedToContactCriteriaBase)> _
		Public Property AssignedTo() As ContactBase
			Get
				Return assignedToCore
			End Get
			Set(ByVal value As ContactBase)
				SetPropertyValue("AssignedTo", assignedToCore, value)
			End Set
		End Property
		<RuleFromBoolProperty(Nothing, DefaultContexts.Save, UsedProperties := "AssignedTo", CustomMessageTemplate := "The type of assigned contact should correspond to the type of the document."), Browsable(False)> _
		Public MustOverride ReadOnly Property AssignedToValidationRuleProperty() As Boolean
	End Class
	Public Class DocumentType1
		Inherits DocumentBase
		Protected Const AssignedToContactCriteria1 As String = "ObjectType.TypeName='WinWebSolution.Module.ContactType1'"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private documentType1PropertyCore As String
		Public Property DocumentType1Property() As String
			Get
				Return documentType1PropertyCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("DocumentType1Property", documentType1PropertyCore, value)
			End Set
		End Property
		<System.ComponentModel.DisplayName("Document Of Type 1 Name")> _
		Public Shadows Property Name() As String
			Get
				Return MyBase.Name
			End Get
			Set(ByVal value As String)
				MyBase.Name = value
			End Set
		End Property
		<DataSourceCriteria(AssignedToContactCriteriaBase & " AND " & AssignedToContactCriteria1)> _
		Public Shadows Property AssignedTo() As ContactBase
			Get
				Return MyBase.AssignedTo
			End Get
			Set(ByVal value As ContactBase)
				MyBase.AssignedTo = value
			End Set
		End Property
		Public Overrides ReadOnly Property AssignedToValidationRuleProperty() As Boolean
			Get
				If AssignedTo IsNot Nothing Then
					Return AssignedTo.GetType() Is GetType(ContactType1)
				End If
				Return True
			End Get
		End Property
	End Class
	Public Class DocumentType2
		Inherits DocumentBase
		Protected Const AssignedToContactCriteria2 As String = "ObjectType.TypeName='WinWebSolution.Module.ContactType2'"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private documentType2PropertyCore As String
		Public Property DocumentType2Property() As String
			Get
				Return documentType2PropertyCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("DocumentType2Property", documentType2PropertyCore, value)
			End Set
		End Property
		<System.ComponentModel.DisplayName("Document Of Type 2 Name")> _
		Public Shadows Property Name() As String
			Get
				Return MyBase.Name
			End Get
			Set(ByVal value As String)
				MyBase.Name = value
			End Set
		End Property
		<DataSourceCriteria(AssignedToContactCriteriaBase & " AND " & AssignedToContactCriteria2)> _
		Public Shadows Property AssignedTo() As ContactBase
			Get
				Return MyBase.AssignedTo
			End Get
			Set(ByVal value As ContactBase)
				MyBase.AssignedTo = value
			End Set
		End Property
		<RuleRequiredField(Nothing, DefaultContexts.Save), VisibleInDetailView(True), VisibleInListView(True), VisibleInLookupListView(True)> _
		Public Shadows Property Author() As String
			Get
				Return MyBase.Author
			End Get
			Set(ByVal value As String)
				MyBase.Author = value
			End Set
		End Property
		Public Overrides ReadOnly Property AssignedToValidationRuleProperty() As Boolean
			Get
				If AssignedTo IsNot Nothing Then
					Return AssignedTo.GetType() Is GetType(ContactType2)
				End If
				Return True
			End Get
		End Property
	End Class
End Namespace

File: CS/WinWebSolution.Module/Documents.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;

namespace WinWebSolution.Module {
    [DefaultClassOptions]
    [System.ComponentModel.DisplayName("Documents")]
    [DefaultProperty("Name")]
    public abstract class DocumentBase : Note {
        protected const string AssignedToContactCriteriaBase = "Active=True";
        public DocumentBase(Session session) : base(session) { }
        [VisibleInDetailView(false)]
        [VisibleInListView(false)]
        [VisibleInLookupListView(false)]
        public new string Author {
            get { return base.Author; }
            set { base.Author = value; }
        }
        private string nameCore;
        [RuleRequiredField(null, DefaultContexts.Save)]
        public string Name {
            get { return nameCore; }
            set { SetPropertyValue("Name", ref nameCore, value); }
        }
        private ContactBase assignedToCore;
        [DataSourceCriteria(AssignedToContactCriteriaBase)]
        public ContactBase AssignedTo {
            get { return assignedToCore; }
            set { SetPropertyValue("AssignedTo", ref assignedToCore, value); }
        }
        [RuleFromBoolProperty(null, DefaultContexts.Save, UsedProperties = "AssignedTo", CustomMessageTemplate = "The type of assigned contact should correspond to the type of the document."), Browsable(false)]
        public abstract bool AssignedToValidationRuleProperty { get;}
    }
    public class DocumentType1 : DocumentBase {
        protected const string AssignedToContactCriteria1 = "ObjectType.TypeName='WinWebSolution.Module.ContactType1'";
        public DocumentType1(Session session) : base(session) { }
        private string documentType1PropertyCore;
        public string DocumentType1Property {
            get { return documentType1PropertyCore; }
            set { SetPropertyValue("DocumentType1Property", ref documentType1PropertyCore, value); }
        }
        [System.ComponentModel.DisplayName("Document Of Type 1 Name")]
        public new string Name {
            get { return base.Name; }
            set { base.Name = value; }
        }
        [DataSourceCriteria(AssignedToContactCriteriaBase + " AND " + AssignedToContactCriteria1)]
        public new ContactBase AssignedTo {
            get { return base.AssignedTo; }
            set { base.AssignedTo = value; }
        }
        public override bool AssignedToValidationRuleProperty {
            get {
                if (AssignedTo != null) {
                    return AssignedTo.GetType() == typeof(ContactType1);
                }
                return true;
            }
        }
    }
    public class DocumentType2 : DocumentBase {
        protected const string AssignedToContactCriteria2 = "ObjectType.TypeName='WinWebSolution.Module.ContactType2'";
        public DocumentType2(Session session) : base(session) { }
        private string documentType2PropertyCore;
        public string DocumentType2Property {
            get { return documentType2PropertyCore; }
            set { SetPropertyValue("DocumentType2Property", ref documentType2PropertyCore, value); }
        }
        [System.ComponentModel.DisplayName("Document Of Type 2 Name")]
        public new string Name {
            get { return base.Name; }
            set { base.Name = value; }
        }
        [DataSourceCriteria(AssignedToContactCriteriaBase + " AND " + AssignedToContactCriteria2)]
        public new ContactBase AssignedTo {
            get { return base.AssignedTo; }
            set { base.AssignedTo = value; }
        }
        [RuleRequiredField(null, DefaultContexts.Save)]
        [VisibleInDetailView(true)]
        [VisibleInListView(true)]
        [VisibleInLookupListView(true)]
        public new string Author {
            get { return base.Author; }
            set { base.Author = value; }
        }
        public override bool AssignedToValidationRuleProperty {
            get {
                if (AssignedTo != null) {
                    return AssignedTo.GetType() == typeof(ContactType2);
                }
                return true;
            }
        }
    }
}